src/mapreduce.jl

```diff
@@ -1,430 +1,214 @@
-#########
-## map ##
-#########
-
-# Single input
-@generated function map{T}(f, a1::StaticArray{T})
-    exprs = [:(f(a1[$j])) for j = 1:length(a1)]
-    return quote
-        $(Expr(:meta, :inline))
-        newtype = similar_type(typeof(a1), promote_op(f, T))
-        @inbounds return $(Expr(:call, :newtype, Expr(:tuple, exprs...)))
+# Returns the common Size of the inputs (or else throws a DimensionMismatch)
+@inline same_size(a1::StaticArray, as::StaticArray...) = _same_size(Size(a1), as...)
+@inline _same_size(s::Size) = s
+@inline function _same_size(s::Size, a1::StaticArray, as::StaticArray...)
+    if s === Size(a1)
+        return _same_size(s, as...)
+    else
+        throw(DimensionMismatch("Dimensions must match. Got inputs with $s and $(Size(a1))."))
     end
 end
 
-# Two inputs
-@generated function map{T1,T2}(f, a1::StaticArray{T1}, a2::StaticArray{T2})
-    if size(a1) != size(a2)
-        error("Dimensions must match. Got sizes $(size(a1)) and $(size(a2))")
-    end
+@inline _first(a1, as...) = a1
 
-    exprs = [:(f(a1[$j], a2[$j])) for j = 1:length(a1)]
-    return quote
-        $(Expr(:meta, :inline))
-        newtype = similar_type(typeof(a1), promote_op(f, T1, T2))
-        @inbounds return $(Expr(:call, :newtype, Expr(:tuple, exprs...)))
-    end
-end
-
-# TODO these assume linear fast...
-@generated function map{T1,T2}(f, a1::StaticArray{T1}, a2::AbstractArray{T2})
-    exprs = [:(f(a1[$j], a2[$j])) for j = 1:length(a1)]
-    return quote
-        $(Expr(:meta, :inline))
-
-        if size(a1) != size(a2)
-            error("Dimensions must match. Got sizes $(size(a1)) and $(size(a2))")
-        end
+################
+## map / map! ##
+################
 
-        newtype = similar_type(typeof(a1), promote_op(f, T1, T2))
-        @inbounds return $(Expr(:call, :newtype, Expr(:tuple, exprs...)))
-    end
+@inline function map(f, a::StaticArray...)
+    _map(f, same_size(a...), a...)
 end
 
-@generated function map{T1,T2}(f, a1::AbstractArray{T1}, a2::StaticArray{T2})
-    exprs = [:(f(a1[$j], a2[$j])) for j = 1:length(a2)]
+@generated function _map(f, ::Size{S}, a::StaticArray...) where {S}
+    exprs = Vector{Expr}(prod(S))
+    for i ∈ 1:prod(S)
+        tmp = [:(a[$j][$i]) for j ∈ 1:length(a)]
+        exprs[i] = :(f($(tmp...)))
+    end
+    eltypes = [eltype(a[j]) for j ∈ 1:length(a)] # presumably, `eltype` is "hyperpure"?
```

Comment:
lol.

Yes - should I make a PR to Julia NEWS.md highlighting our new *hyperpure*™ technology stack?